src/tool/abi/abi_writer.h

```diff
@@ -0,0 +1,251 @@
+#pragma once
+
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include "common.h"
+#include "meta_reader.h"
+#include "metadata_cache.h"
+#include "namespace_iterator.h"
+#include "text_writer.h"
+
+struct basic_writer : xlang::text::writer_base<basic_writer> {};
+
+template <typename Int>
+struct format_hex
+{
+    static_assert(std::is_unsigned_v<Int>);
+    Int value;
+};
+template <typename Int>
+format_hex(Int) -> format_hex<Int>;
+
+struct indent { std::size_t additional_indentation = 0; };
+
+struct writer : xlang::text::writer_base<writer>
+{
+    writer(abi_configuration const& config) :
+        m_config(config)
+    {
+    }
+
+    using writer_base::write;
+
+    abi_configuration const& config() const noexcept
+    {
+        return m_config;
+    }
+
+    void write_code(std::string_view value)
+    {
+        xlang::text::bind_list("::", namespace_range{ value })(*this);
+    }
+
+    static void write_uppercase(writer& w, std::string_view str)
```

Comment:
write_uppercase and write_lowercase run the risk of misbehaving with non-ASCII characters. @kennykerr we should add some types to our test component with names like Çömpôñénť.

Reply:
That's a good point, though MIDL is currently a very unhappy camper when I try and use non-ANSI characters

Reply:
Changed characters and got it to compile and output a .winmd and .h (though, with an assertion failure), but the .h is bogus...